docs/notebooks/Poincare Evaluation.ipynb

```diff
@@ -0,0 +1,1448 @@
+    "    col_lengths = [len(col) for col in data_cols]\n",
+    "    header_col_1_length = col_lengths[0] + col_lengths[1] - 1\n",
+    "    header_col_2_length = sum(col_lengths[2:]) + len(col_lengths[2:-1]) - 2\n",
+    "    \n",
+    "    header_col_2_content = \"Model Dimensions\"\n",
+    "    header_col_2_left_margin = (header_col_2_length - len(header_col_2_content)) // 2\n",
+    "    header_col_2_right_margin = header_col_2_length - len(header_col_2_content) - header_col_2_left_margin\n",
+    "    header_col_2_string = \"%s%s%s\" % (\n",
+    "        \" \" * header_col_2_left_margin, header_col_2_content, \" \" * header_col_2_right_margin)\n",
+    "\n",
+    "    header = PrettyTable()\n",
+    "    header.field_names = [\" \" * header_col_1_length, header_col_2_string]\n",
+    "    header_lines = header.get_string(start=0, end=0).split(\"\\n\")[:2]\n",
+    "    print('Results for %s task' % task_name)\n",
+    "    print(\"\\n\".join(header_lines))\n",
+    "    print(data)        "
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "### 4.1 WordNet reconstruction"
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": 34,
+   "metadata": {},
+   "outputs": [],
+   "source": [
+    "import csv\n",
+    "from collections import defaultdict\n",
+    "import itertools\n",
+    "\n",
+    "\n",
+    "class ReconstructionEvaluation(object):\n",
+    "    \"\"\"Evaluating reconstruction on given network for given embedding\"\"\"\n",
+    "    def __init__(self, filepath, embedding):\n",
+    "        \"\"\"Initialize evaluation instance with tsv file containing relation pairs and embedding to be evaluated\n",
+    "        \n",
+    "        Args:\n",
+    "            filepath (str): path to tsv file containing relation pairs\n",
+    "            embedding (PoincareEmbedding instance): embedding to be evaluated\n",
+    "        \n",
+    "        Returns\n",
+    "            ReconstructionEvaluation instance\n",
+    "\n",
+    "        \"\"\"\n",
+    "        items = set()\n",
+    "        embedding_vocab = embedding.kv.vocab\n",
+    "        relations = defaultdict(set)\n",
+    "        with smart_open(filepath, 'r') as f:\n",
+    "            reader = csv.reader(f, delimiter='\\t')\n",
+    "            for row in reader:\n",
+    "                assert len(row) == 2, 'Hypernym pair has more than two items'\n",
+    "                item_1_index = embedding_vocab[row[0]].index\n",
+    "                item_2_index = embedding_vocab[row[1]].index\n",
+    "                relations[item_1_index].add(item_2_index)\n",
+    "                items.update([item_1_index, item_2_index])\n",
+    "        self.items = items\n",
+    "        self.relations = relations\n",
+    "        self.embedding = embedding\n",
+    "    \n",
+    "    \n",
+    "    @staticmethod\n",
+    "    def get_positive_relation_ranks(distances, positive_relations):\n",
```

Comment:
There's a bit too much magic in this method implementation. Are you sure it actually does what it's expected to do? Have you tested it?

Reply:
Yeah I agree the code isn't straightforward. I changed `distances` to `all_distances` to make it slightly clearer. And yes, I've tested it - 
```python
all_distances = np.array([5, 10, 15, 20, 0, 1, 2, 3])
positive_relations = [4, 6, 1, 7]
assert(ReconstructionEvaluation.get_positive_relation_ranks(all_distances, positive_relations) == [1, 2, 3, 2])
```

Reply:
My concern wasn't about naming, but rather about the used constructions. Mainly
```python
ranks = (negative_relation_distances < positive_relation_distances[:, np.newaxis]).sum(axis=1) + 1
```
is pretty non-obvious. I have tested it myself, though, and indeed does seem to work, so that's good.